app/app_main.c

```diff
@@ -1678,6 +1699,129 @@ static ACVP_RESULT app_hmac_handler(ACVP_TEST_CASE *test_case)
     return ACVP_SUCCESS;
 }
 
+static ACVP_RESULT app_kdf135_tls_handler(ACVP_TEST_CASE *test_case)
+{
+    ACVP_KDF135_TLS_TC	*tc;
+    unsigned char *key_block1, *key_block2, *master_secret1, *master_secret2;
+    int olen1 = 0, olen2 = 0, len1, ret, i, len, count, psm_len;
+    const EVP_MD *evp_md1 = NULL, *evp_md2 = NULL;
+
+    tc = test_case->tc.kdf135_tls;
+    /* We only support TLS12 for now */
+    if (tc->method != ACVP_KDF135_TLS12) {
+        printf("\nCrypto module error, Bad TLS type\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    olen1 = tc->pm_len;
+    olen2 = tc->kb_len;
+    key_block1 = tc->kblock1;
+    key_block2 = tc->kblock2;
+    master_secret1 = tc->msecret1;
+    master_secret2 = tc->msecret2;
+
+    if (!key_block1 || !key_block2 || !master_secret1 || !master_secret2) {
+        printf("\nCrypto module error, malloc failure\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    switch (tc->md) 
+    {
+    case ACVP_KDF135_TLS_CAP_SHA256:
+        evp_md1 = evp_md2 = EVP_sha256();
+	break;
+    case ACVP_KDF135_TLS_CAP_SHA384:
+        evp_md1 = evp_md2 = EVP_sha384();
+	break;
+    case ACVP_KDF135_TLS_CAP_SHA512:
+        evp_md1 = evp_md2 = EVP_sha512();
+	break;
+    default:
+        printf("\nCrypto module error, Bad SHA type\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    count = 1;
+    len = tc->pm_len / count;
+    if (count == 1)
+          psm_len = 0;
+
+    memset(master_secret1, 0, 4096);
+ 
+    ret = kdf_tls12_P_hash(evp_md1, (const unsigned char *)tc->pm_secret, len + (psm_len & 1), 
+	                   TLS_MD_MASTER_SECRET_CONST, TLS_MD_MASTER_SECRET_CONST_SIZE,
+			   tc->ch_rnd, strlen((char *)tc->ch_rnd),
+			   tc->sh_rnd, strlen((char *)tc->sh_rnd),
+			   NULL, 0,
+			   NULL, 0,
+			   master_secret1, olen1);
+    if (ret == 0) {
+        printf("\nCrypto module error, TLS kdf failure\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    for (i = 0; i < olen1; i++) {
+         master_secret1[i] ^= master_secret2[i];
+    }    
+
+    if (evp_md1 != evp_md2) {
+        ret = kdf_tls12_P_hash(evp_md2, (const unsigned char *)tc->pm_secret + len, len + (psm_len & 1), 
+	                       TLS_MD_MASTER_SECRET_CONST, TLS_MD_MASTER_SECRET_CONST_SIZE,
+			       tc->ch_rnd, strlen((char *)tc->ch_rnd),
+			       tc->sh_rnd, strlen((char *)tc->sh_rnd),
+			       NULL, 0,
+			       NULL, 0,
+			       master_secret2, olen1);
+	if (ret == 0) {
+            printf("\nCrypto module error, TLS kdf failure\n");
+            return ACVP_CRYPTO_MODULE_FAIL;
+        }
+        for (i = 0; i < olen1; i++) {
+            master_secret1[i] ^= master_secret2[i];
+        }
+    }
+
+
+    memset(key_block1, 0, 4096);
```

Comment:
Good catch, I should clear that in the acvp code not the app code where I already have the size.